README.rst
about
-----

bot-prakasha-ke is an IRC bot which does the following:
 * connects to arbitrary channels on an IRC host
 * logs those channels (topics and conversations)
 * broadcasts messages or changes topics
 * offers a web service view viewing recorded logs
 * offers an ssh shell of the running bot for batch-updating channel topics
   or broadcasting messages
 * allows one to use the ssh command to script broadcast messages, topic
   updates, etc., e.g.r::
     ssh -p 6622 127.0.0.1 "say('#cool-channel', 'Have a good night!')"

quick start
-----------

If you install with pip, the dependencies will be added for you::

  $ pip install .

To run the server with the defaults, simply do this::

  make run

That will log to stdout, so it's nice for debugging. If you want to run as a
daemon::

  make daemon

To stop the daemon::

  make stop


configuration
-------------

A configuration file is automatically generated when you do ``make run`` or
``make daemon``. However, if you'd like to pre-generate the config file so that
you can edit it prior to running the server, you can do this::

  make generate-config

WARNING!

    Running the make target ``make generate-config`` will overwrite your
    existing configuration! Backup your configuration file before using!


ChangeLog
2012-05-12

* Updated TODO.
* Updated setup.py to install DreamSSH as a requirement.

2012-04-25

* Added an SFTP implementation that aims to push files into the running memory
  of the bot.

2012-04-24

* Fixed bug in configuration logic that was setting sslEnabled to true.
* Stubbed out some IRC methods in the publisher client.
* Added API calls for getting the users on a given channel.
* Added missing docstrings in API.
* Updated TODO.
* More TODO updates.

2012-04-23

* Added support for ini files (local dir and user home dir).
* Added a make target for importing bzr into a git checkout.
* Added make target for adding files to both bzr and git.
* Added convenience scripts for starting and stopping the bot server.
* Added a make target for updating git and bzr.
* Updated the README with "getting started" instructions.
* Updated the setup.py with deps.
* Moved the util scripts to make targets and added more make targets.
* Updated config file to convert unicode to str where necessary (for IRC).
* Cleaned up the code for getting the priv and pub keys.
* Updated TODO.
* More tweaks to the Makefile to support non-synced changes to bzr and git
  repos.
* Updated README with notes about configuration.
* Added another feature description.
* Noted a couple bugs.
* Changed the shell prompt.
* Moved the banner function into the API class.
* Added an API call for listing the channels.

2012-04-22

* Update dist util module to use .txt or .rst.
* Changed setup.py to point to the .rst docs.
* Fixed the Makefile to do "check" against the lib name, not the project name.
* Added more TODO items.
* Renamed all internal references of publishbot to prakasha.
* Added an exceptions module.
* Updated the key-getter to raise a custom exception if the server keys aren't
  present (and to give instructions on how to generate them).
* Added support for executing commands via the ssh command, such as:
    $ ssh -p 6622 "say('#adytum', 'wassup?')"
* Added new custom exception for commands that aren't defined in the namespace.
* Added windowChanged method.

2012-04-21

* More TODO changes.
* Started on README changes.
* Updated TODO.
* Updated ignore files.
* Added Hinglish font "bot" images.
* Updated banner in configuration.
* Tweaked Makefile.
* Added a setup.py file.
* Added distutil wrappers.
* Added a meta file.
* Added git ignore file.
* Updated ignores.
* Changed docs to ReST files.
* Added TODO and docs.
* Various admin updates.
* Added another .ignore option as well as a Makefile.
* Renamed README from .txt to .rst.
* Added project logos.

2012-04-17

* More tweaks to the API.
* Fixed bad call to topic.
* Renamed methods.
* Added new methods.
* Added docstrings.
* Added more CommandAPI methods and placed them in the default namespace.
* Added a command api for use inside the publishbot shell (via ssh).
* Converted all services to named services.
* Separated logger and rotator in config and service setup.
* Added support for MOTD or a banner, though there's still a clearscreen
  happening (in order to currently see the banner, you have to scroll up).
* Cleaned up namespace for SSH Python shell.
* Added support for SSH'ing into publishbot.
* Turns out the nick is that of who's logged in, not who actually set the
  topic.
* Added config options for SSH and tweaked port numbers.
* Added more support for displaying what a channel's topic was.
* First attempt at getting topic changes logged.
* Playing around with some shell possibilities. Probably going to delete this
  and use a conch shell.

2012-04-16

* Added a script to generate SSH keys.
* Added some quick scripting for a raw IRC client.


2012-04-11

* Changed the nick of the logger.
* Added a .bzrignore file.
* This one had the most changed of all the old modules. Since this was
  originally written, Twisted has added built-in support for BasicAuth and has
  added the web2 stuff to web (thus changing the old web API, module imports,
  etc.).
* More formatting and another move into the new package.
* Tweaked the formatting and moved the module.
* Moved the config module and updated some of the options.
* Moved the committer code into the new package.
* Created a new package for the codebase.
* Created a bin dir and moved tjhe tac file in there.


2007-09-03

* Ooops; should be 24 hours by default, not 12.
* Added support for basic auth and updated bot.tac accordingly.

2007-09-02

* Removed needless classes from config file.

2007-08-23

* Removed hard-coded 24 hour time check.
* Made time of rotation configurable, removing hard-coded 'midnight' as the
  only option.

2007-08-19

* Removed commented debugging code.
* Fixed math for 'hoursAgo' variable.

2007-08-18

* Split up logger and publisher code into different modules.
* Formatting changes.
* Code shuffling.
* Debgging statements.

2007-08-17

* Changed the directory structure of the logfiles a little bit, so as to be
  more readable.
* Fixed log rotation issues.
* Removed the crappy code that created a client/connection for every channel;
  the code now creates one connection for the irc server.
* Got log rotation partially working.
* Added logging support.
* Added vhost web server support.
* Created a publisher factory and removed the corresponding code from the .tac
* Added log configuration
* Added docstring for listener.
* Redid configuration user use object attrs instead of constants.
* Created branch of radix's publish-bot.

2007-06-13

* use ReconnectingClientFactory for more resilient IRC connections.


2007-05-10

* Oops, README.txt thought the post commit function was "post_commit" but it
  was "send_commit".

2007-03-20

* merge jkakar's error-handling of socket.connect.
* Added try/except block around socket.connect() call. If a connection
  can't be opened a useful message is printed (instead of a traceback).


2007-03-08

* Merge branch-prefix from jamu. Thanks.
* message_branch_prefix configuration parameter helps disambiguate
  which project a branch belongs to.

prakasha/config.py
from ConfigParser import SafeConfigParser
import json
import os

from zope.interface import moduleProvides

from dreamssh.config import Config, Configurator, main, ssh
from dreamssh.sdk import interfaces


moduleProvides(interfaces.IConfig)


# Main
main.config.userdir = os.path.expanduser("~/.prakasha-bot")
main.config.userfile = "%s/%s" % (main.config.userdir, main.config.localfile)

# Internal SSH Server
ssh.servicename = "SSH Server"
ssh.port = 6622
ssh.keydir = ".prakasha-ssh"
ssh.banner = """:
: Welcome to
:
:    .;@@
:      .@@@
:        @@@                                 +'
:        '@@;                                +@@`
:      ::,@@@::::::::::::::::::::::::::::::::#@@::::::::,
:      @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@#
:         @@@                       @@@      +@@`
:         @@@    '#+,        `#@@@#,@@@      +@@`
:         @@@ .@@@@@@@+    `@@@@@'.`:@@      +@@`
:         @@@@.    .@@@@  .@@@@       @@     +@@`
:         @@@        ;@@` @@@'        `@@    +@@`
:         @@@@`       `@  +@`        `@@@    +@@`   @
:         @@@@@:      @    #,       '@@@'    +@@`   @@@
:         @@#+@@@.   @      `+   `#@@@@+      @@.  ,@@.
:         #@#  @@@@@          `#@@@@@#         @@'#@#
:
:                       __            __          __
:        ___  _______ _/ /_____ ____ / /  ___ _  / /_____
:       / _ \/ __/ _ `/  '_/ _ `(_-</ _ \/ _ `/ /  '_/ -_)
:      / .__/_/  \_,_/_/\_\\\\_,_/___/_//_/\_,_/ /_/\_\\\\__/
:     /_/
:
: You are in the prakasha interactive Python shell.
: Type 'dir()' to see the objects in the current namespace.
:
: Enjoy!
:
"""

# IRC
irc = Config()
irc.servicename = "IRC Client"
irc.nick = 'prakasha-bot'
irc.server = 'irc.freenode.net'
irc.port = 6667
irc.serverPassword = None
irc.sslEnabled = False
irc.lineRate = 1

# Listener
listener = Config()
listener.servicename = "Message Service"
listener.host = "127.0.0.1"
listener.port = 6666
listener.password = 'ircb0tz'

# Log
log = Config()
log.servicename = "Logger Service"
log.nick = 'prakasha-log'
log.channels = ["#adytum", "#uls"]

# Log rotator
log.rotate = Config()
log.rotate.servicename = "Log Rotator Service"
log.rotate.checkInterval = 60
log.rotate.time = '00:00'
log.rotate.maxAgeHours = 24

# Log web server
log.http = Config()
log.http.servicename = "Log Browser"
log.http.port = 6680
log.http.docRoot = 'irclogs'
log.http.vhostEnabled = True
log.http.vhost= 'localhost'
#log.http.auth = 'basic'
log.http.auth = None
log.http.realm = 'bot-prakasha-ke IRC Logs'

# Log web server users
log.http.users = {
    'Jojo': 'c1rcu5b0y',
    }
log.http.userdb = "httpd.password"


class PrakashaConfigurator(Configurator):

    def __init__(self, main, ssh, irc, listener, log):
        super(PrakashaConfigurator, self).__init__(main, ssh)
        self.irc = irc
        self.listener = listener
        self.log = log

    def buildDefaults(self):
        config = super(PrakashaConfigurator, self).buildDefaults()
        config.add_section("IRC")
        config.set("IRC", "servicename", self.irc.servicename)
        config.set("IRC", "nick", self.irc.nick)
        config.set("IRC", "server", self.irc.server)
        config.set("IRC", "port", str(self.irc.port))
        config.set("IRC", "serverPassword", self.irc.serverPassword or "")
        config.set("IRC", "sslEnabled", str(self.irc.sslEnabled))
        config.set("IRC", "lineRate", str(self.irc.lineRate))

        config.add_section("Listener")
        config.set("Listener", "servicename", self.listener.servicename)
        config.set("Listener", "host", self.listener.host)
        config.set("Listener", "port", str(self.listener.port))
        config.set("Listener", "password", self.listener.password)

        config.add_section("Logger")
        config.set("Logger", "servicename", self.log.servicename)
        config.set("Logger", "nick", self.log.nick)
        config.set("Logger", "channels", json.dumps(self.log.channels))

        config.add_section("Rotator")
        config.set("Rotator", "servicename", self.log.rotate.servicename)
        config.set(
            "Rotator", "checkInterval", str(self.log.rotate.checkInterval))
        config.set("Rotator", "time", self.log.rotate.time)
        config.set("Rotator", "maxAgeHours", str(self.log.rotate.maxAgeHours))

        config.add_section("HTTP")
        config.set("HTTP", "servicename", self.log.http.servicename)
        config.set("HTTP", "port", str(self.log.http.port))
        config.set("HTTP", "docRoot", self.log.http.docRoot)
        config.set("HTTP", "vhostEnabled", str(self.log.http.vhostEnabled))
        config.set("HTTP", "vhost", self.log.http.vhost)
        config.set("HTTP", "auth", str(self.log.http.auth))
        config.set("HTTP", "realm", self.log.http.realm)
        config.set("HTTP", "users", json.dumps(self.log.http.users))
        config.set("HTTP", "userdb", self.log.http.userdb)
        return config

    def updateConfig(self):
        config = super(PrakashaConfigurator, self).updateConfig()
        irc = self.irc
        listener = self.listener
        log = self.log
        http = self.log.http
        rotate = self.log.rotate

        # IRC
        irc.servicename = config.get("IRC", "servicename")
        irc.nick = str(config.get("IRC", "nick"))
        irc.server = str(config.get("IRC", "server"))
        irc.port = int(config.get("IRC", "port"))
        irc.serverPassword = config.get("IRC", "serverPassword") or None
        sslEnabled = config.get("IRC", "sslEnabled")
        if sslEnabled.lower() == "true":
            sslEnabled = True
        else:
            sslEnabled = False
        irc.sslEnabled = sslEnabled
        irc.lineRate = int(config.get("IRC", "lineRate"))

        # Listener
        listener.servicename = config.get("Listener", "servicename")
        listener.host = config.get("Listener", "host")
        listener.port = int(config.get("Listener", "port"))
        listener.password = config.get("Listener", "password")

        # Log
        log.servicename = config.get("Logger", "servicename")
        log.nick = str(config.get("Logger", "nick"))
        log.channels = [str(x) for x in
                        json.loads(config.get("Logger", "channels"))]

        # Log rotator
        rotate.servicename = config.get("Rotator", "servicename")
        rotate.checkInterval = int(config.get("Rotator", "checkInterval"))
        rotate.time = config.get("Rotator", "time") 
        rotate.maxAgeHours = int(config.get("Rotator", "maxAgeHours"))

        # Log web server
        http.servicename = config.get("HTTP", "servicename")
        http.port = int(config.get("HTTP", "port"))
        http.docRoot = config.get("HTTP", "docRoot")
        http.vhostEnabled = bool(config.get("HTTP", "vhostEnabled"))
        http.vhost= config.get("HTTP", "vhost")
        http.auth = config.get("HTTP", "auth")
        http.realm = config.get("HTTP", "realm")

        # Log web server users
        http.users = json.loads(config.get("HTTP", "users"))
        http.userdb = config.get("HTTP", "userdb")

        return config


configurator = PrakashaConfigurator(main, ssh, irc, listener, log)
configurator.updateConfig()
